lib/spack/llnl/util/filesystem.py

```diff
@@ -693,18 +698,24 @@ def replace_directory_transaction(directory_name, tmp_root=None):
     try:
         yield tmp_dir
     except (Exception, KeyboardInterrupt, SystemExit):
-        # Delete what was there, before copying back the original content
-        if os.path.exists(directory_name):
-            shutil.rmtree(directory_name)
-        shutil.move(
-            src=os.path.join(tmp_dir, directory_basename),
-            dst=os.path.dirname(directory_name)
-        )
+        # Try to recover the original directory, if this fails, raise a
+        # composite exception.
+        try:
+            # Delete what was there, before copying back the original content
+            if os.path.exists(directory_name):
+                shutil.rmtree(directory_name)
+            shutil.move(
+                src=os.path.join(tmp_dir, directory_basename),
+                dst=os.path.dirname(directory_name)
+            )
+        except OSError as e:
```

Comment:
A few things just occurred to me that might be worth addressing before merging this:

* It might be worth knowing the original exception that led us to the `try:` in the first place: at the moment it is lost
* Are there other types of exceptions that might occur here? Should we catch all of them that are of type `Exception`?
* Would it be possible to add a test? Some combination of using the `canfail` package and monkey-patching `shutil.rmtree` to fail could work

Reply:
@scheibelp I don't like the idea of monkeypatching such a common function, please check the last commit for an attempt to unit test this. If there's a simpler way to get a 'build task' without instantiating the installer itself, I'm all ears.

The first two points are addressed too now